src/modal.js

```diff
@@ -0,0 +1,209 @@
+/**
+ * Given a target element, finds the Elements to hide for accessibility.
+ * @param {!Element} element The target Element.
+ * @return {!Array<!Element>} The Elements to add hide from accessibiluty.
+ * @package Visible for testing
+ */
+export function getElementsToAriaHide(element) {
+  const arr = [];
+  const ancestors = getAncestors(element);
+
+  for (let i = 0; i < ancestors.length; i++) {
+    const cur = ancestors[i];
+
+    if (!cur.parentNode) {
+      continue;
+    }
+
+    toArray(cur.parentNode.children)
+      .filter(c => c != cur)
+      .forEach(c => arr.push(c));
+  }
+
+  return arr;
+}
+
+/**
+ * Gets all ancestors of an Element, including ancestors of any ShadowRoot
+ * hosts.
+ * @param {!Element} element
+ * @return {!Array<!Element>}
+ */
+function getAncestors(element) {
+  const ancestry = [];
+
+  for (let cur = element; cur; cur = cur.parentNode || cur.host) {
+    ancestry.push(cur);
+  }
+
+  return ancestry;
+}
+
+/**
+ * Gets the potentially focusable elements, assuming something within the
+ * Element has focus. Note that we do not need to go into ShadowRoots, giving
+ * their host `tabindex="-1"` is sufficient.
+ *
+ * Note that some of these Elements may not be focusable (e.g. is a button
+ * that is `disabled` or has an ancestor that is `display: none`).
+ * @param {!Element} element
+ * @return {!Array<!Element>}
+ */
+function getPotentiallyFocusableElements(element) {
+  const arr = [];
+  let cur = element;
+
+  while (cur) {
+    const root = rootNodeFor(cur);
+    const potentiallyFocusable = root.querySelectorAll(
+      [
+        'a[href]',
+        'area[href]',
+        'button',
+        'details',
+        'iframe',
+        'input',
+        'select',
+        'textarea',
+        '[contenteditable]',
+        '[tabindex]',
+      ].join(',')
+    );
+    Array.prototype.push.apply(arr, potentiallyFocusable);
+
+    cur = root.host;
+  }
+
+  return arr.filter(e => !element.contains(e));
+}
+
+/**
+ * Sets an Element as an open modal, making all Elements outside of the page
+ * hidden from the tab order and screenreaders.
+ *
+ * Note: this does not block click events on things outside of the modal. It is
+ * assumed that a backdrop Element blocking clicks is present.
+ * @param {!Element} element
+ */
+export function setModalAsOpen(element) {
+  devAssert(modalEntryStack.every(info => info.element !== element));
+  devAssert(isConnectedNode(element));
+
+  const elements = getElementsToAriaHide(element);
+  const ancestry = getAncestors(element).filter(
+    n => n.nodeType == Node.ELEMENT_NODE
+  );
+  const focusableElements = getPotentiallyFocusableElements(element);
+  const hiddenElementInfos = elements.concat(ancestry).map(element => {
+    return {
+      element,
+      prevValue: element.getAttribute('aria-hidden'),
+    };
+  });
+  const focusableElementInfos = focusableElements.map(element => {
+    return {
+      element,
+      prevValue: element.getAttribute('tabindex'),
```

Comment:
If two modals are opened, I think this will create `N` more `{element,prevValue}` objects with elements previously assigned `tabindex="-1"`. Should we try to avoid setting the `tabindex` to `-1` for elements that are already `tabindex="-1"`? I can't tell if that would be a useful optimization or not.

Reply:
Maybe a test for nested modals checking that `tabindex` is updated appropriately would be sufficient, like you have for `aria-hidden`